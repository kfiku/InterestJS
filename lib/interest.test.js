/* eslint-env jest */
'use strict'
var assert = require('assert')
var Interest = require('../lib/interest.js')

describe('interestjs node module.', function () {
  it('must count interest correctly', function () {
    var interest = new Interest(1000, 12, 5, { tax: 19 })
    var taxSum = 0
    var interestSum = 0
    var p
    var i = 0

    assert.strictEqual(interest.payments.length, 12, 'payments are not equal to 120')
    for (i; i < interest.payments.length; i++) {
      p = interest.payments[i]
      assert(p.capital > 0, true)
      assert(p.interest > 0, true)
      assert(p.tax > 0, true)

      taxSum += p.tax
      interestSum += p.interest
    }

    assert(interest.interestSum > 0, true, 'wrong interestSum')
    assert.strictEqual(interest.interestSum, interestSum,
      'interestsSum (' + interest.interestSum + ') not exual to sum of all interests (' + interestSum + ') in payments array')
    assert.strictEqual(interest.taxSum, taxSum,
      'taxSum (' + interest.taxSum + ') not exual to taxSum of all taxes (' + taxSum + ') in payments array')
  })

  it('must count interest correctly for dynamic amount', function () {
    var interest = new Interest(1000, 12, 5, {
      tax: 19,
      dynamicAmount: function (i) {
        return 100 + i * 10
      }
    })

    var taxSum = 0
    var interestSum = 0
    var p
    var i = 0

    assert.strictEqual(interest.payments.length, 12, 'payments are not equal to 12')
    for (i; i < interest.payments.length; i++) {
      p = interest.payments[i]
      assert(p.capital > 0, true)
      assert(p.interest > 0, true)
      assert(p.tax > 0, true)

      taxSum += p.tax
      interestSum += p.interest
    }

    assert(interest.interestSum > 0, true, 'wrong interestSum')
    assert.strictEqual(interest.interestSum, interestSum,
      'interestsSum (' + interest.interestSum + ') not exual to sum of all interests (' + interestSum + ') in payments array')
    assert.strictEqual(interest.taxSum, taxSum,
      'taxSum (' + interest.taxSum + ') not exual to taxSum of all taxes (' + taxSum + ') in payments array')
  })

  it('should count without params', function () {
    const interest = new Interest(1000, 12, 5)
    expect(interest.sum).toBe(12330.01)
    expect(interest).toMatchSnapshot()
  })

  it('should count with startAmount param', function () {
    const interest = new Interest(1000, 12, 5, { startAmount: 1000 })
    expect(interest.sum).toBe(13381.16)
    expect(interest).toMatchSnapshot()
  })

  it('should count with tax param', function () {
    const interest = new Interest(1000, 12, 5, { tax: 19 })
    expect(interest.sum).toBe(12266.53)
    expect(interest).toMatchSnapshot()
  })

  it('should count with tax param', function () {
    const interest = new Interest(1000, 12, 5, { contrib: 100 })
    expect(interest.sum).toBe(13563.03)
    expect(interest).toMatchSnapshot()
  })

  it('should count with dynamicAmount param', function () {
    const dynamicAmount = (i) => {
      return 100 + i * 10
    }
    const interest = new Interest(1000, 12, 5, { dynamicAmount })
    expect(interest.sum).toBe(1905.04)
    expect(interest).toMatchSnapshot()
  })

  it('shout throw on wrong params', function () {
    expect(() => Interest(20000, 'asd')).toThrowError('wrong parameters (20000, "asd", undefined, undefined)')
    expect(() => Interest(1000, 12, '5')).toThrowError('wrong parameters (1000, 12, "5", undefined)')
    expect(() => Interest(1000, 12, 5, true)).toThrowError('wrong parameters (1000, 12, 5, true)')
  })
})
