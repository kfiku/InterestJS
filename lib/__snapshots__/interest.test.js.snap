// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interestjs node module. should count with dynamicAmount param 1`] = `
Object {
  "capitalSum": 1860,
  "contributions": 0,
  "interestSum": 45.04,
  "payments": Array [
    Object {
      "capital": 100,
      "capitalSum": 100,
      "contrib": 0,
      "interest": 0.42,
      "sum": 100.42,
      "tax": 0,
    },
    Object {
      "capital": 110,
      "capitalSum": 210,
      "contrib": 0,
      "interest": 0.88,
      "sum": 211.3,
      "tax": 0,
    },
    Object {
      "capital": 120,
      "capitalSum": 330,
      "contrib": 0,
      "interest": 1.38,
      "sum": 332.68,
      "tax": 0,
    },
    Object {
      "capital": 130,
      "capitalSum": 460,
      "contrib": 0,
      "interest": 1.93,
      "sum": 464.61,
      "tax": 0,
    },
    Object {
      "capital": 140,
      "capitalSum": 600,
      "contrib": 0,
      "interest": 2.52,
      "sum": 607.13,
      "tax": 0,
    },
    Object {
      "capital": 150,
      "capitalSum": 750,
      "contrib": 0,
      "interest": 3.15,
      "sum": 760.28,
      "tax": 0,
    },
    Object {
      "capital": 160,
      "capitalSum": 910,
      "contrib": 0,
      "interest": 3.83,
      "sum": 924.11,
      "tax": 0,
    },
    Object {
      "capital": 170,
      "capitalSum": 1080,
      "contrib": 0,
      "interest": 4.56,
      "sum": 1098.67,
      "tax": 0,
    },
    Object {
      "capital": 180,
      "capitalSum": 1260,
      "contrib": 0,
      "interest": 5.33,
      "sum": 1284,
      "tax": 0,
    },
    Object {
      "capital": 190,
      "capitalSum": 1450,
      "contrib": 0,
      "interest": 6.14,
      "sum": 1480.14,
      "tax": 0,
    },
    Object {
      "capital": 200,
      "capitalSum": 1650,
      "contrib": 0,
      "interest": 7,
      "sum": 1687.14,
      "tax": 0,
    },
    Object {
      "capital": 210,
      "capitalSum": 1860,
      "contrib": 0,
      "interest": 7.9,
      "sum": 1905.04,
      "tax": 0,
    },
  ],
  "sum": 1905.04,
  "taxSum": 0,
}
`;

exports[`interestjs node module. should count with startAmount param 1`] = `
Object {
  "capitalSum": 12000,
  "contributions": 0,
  "interestSum": 381.16,
  "payments": Array [
    Object {
      "capital": 1000,
      "capitalSum": 1000,
      "contrib": 0,
      "interest": 8.33,
      "sum": 2008.33,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 2000,
      "contrib": 0,
      "interest": 12.53,
      "sum": 3020.86,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 3000,
      "contrib": 0,
      "interest": 16.75,
      "sum": 4037.61,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 4000,
      "contrib": 0,
      "interest": 20.99,
      "sum": 5058.6,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 5000,
      "contrib": 0,
      "interest": 25.24,
      "sum": 6083.84,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 6000,
      "contrib": 0,
      "interest": 29.52,
      "sum": 7113.36,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 7000,
      "contrib": 0,
      "interest": 33.81,
      "sum": 8147.17,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 8000,
      "contrib": 0,
      "interest": 38.11,
      "sum": 9185.28,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 9000,
      "contrib": 0,
      "interest": 42.44,
      "sum": 10227.72,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 10000,
      "contrib": 0,
      "interest": 46.78,
      "sum": 11274.5,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 11000,
      "contrib": 0,
      "interest": 51.14,
      "sum": 12325.64,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 12000,
      "contrib": 0,
      "interest": 55.52,
      "sum": 13381.16,
      "tax": 0,
    },
  ],
  "sum": 13381.16,
  "taxSum": 0,
}
`;

exports[`interestjs node module. should count with tax param 1`] = `
Object {
  "capitalSum": 12000,
  "contributions": 0,
  "interestSum": 329.05,
  "payments": Array [
    Object {
      "capital": 1000,
      "capitalSum": 1000,
      "contrib": 0,
      "interest": 4.17,
      "sum": 1003.38,
      "tax": 0.79,
    },
    Object {
      "capital": 1000,
      "capitalSum": 2000,
      "contrib": 0,
      "interest": 8.35,
      "sum": 2010.14,
      "tax": 1.59,
    },
    Object {
      "capital": 1000,
      "capitalSum": 3000,
      "contrib": 0,
      "interest": 12.54,
      "sum": 3020.3,
      "tax": 2.38,
    },
    Object {
      "capital": 1000,
      "capitalSum": 4000,
      "contrib": 0,
      "interest": 16.75,
      "sum": 4033.87,
      "tax": 3.18,
    },
    Object {
      "capital": 1000,
      "capitalSum": 5000,
      "contrib": 0,
      "interest": 20.97,
      "sum": 5050.86,
      "tax": 3.98,
    },
    Object {
      "capital": 1000,
      "capitalSum": 6000,
      "contrib": 0,
      "interest": 25.21,
      "sum": 6071.28,
      "tax": 4.79,
    },
    Object {
      "capital": 1000,
      "capitalSum": 7000,
      "contrib": 0,
      "interest": 29.46,
      "sum": 7095.14,
      "tax": 5.6,
    },
    Object {
      "capital": 1000,
      "capitalSum": 8000,
      "contrib": 0,
      "interest": 33.73,
      "sum": 8122.46,
      "tax": 6.41,
    },
    Object {
      "capital": 1000,
      "capitalSum": 9000,
      "contrib": 0,
      "interest": 38.01,
      "sum": 9153.25,
      "tax": 7.22,
    },
    Object {
      "capital": 1000,
      "capitalSum": 10000,
      "contrib": 0,
      "interest": 42.31,
      "sum": 10187.52,
      "tax": 8.04,
    },
    Object {
      "capital": 1000,
      "capitalSum": 11000,
      "contrib": 0,
      "interest": 46.61,
      "sum": 11225.27,
      "tax": 8.86,
    },
    Object {
      "capital": 1000,
      "capitalSum": 12000,
      "contrib": 0,
      "interest": 50.94,
      "sum": 12266.53,
      "tax": 9.68,
    },
  ],
  "sum": 12266.53,
  "taxSum": 62.52,
}
`;

exports[`interestjs node module. should count with tax param 2`] = `
Object {
  "capitalSum": 13200,
  "contributions": 1200,
  "interestSum": 363.03,
  "payments": Array [
    Object {
      "capital": 1000,
      "capitalSum": 1100,
      "contrib": 100,
      "interest": 4.58,
      "sum": 1104.58,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 2200,
      "contrib": 100,
      "interest": 9.19,
      "sum": 2213.77,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 3300,
      "contrib": 100,
      "interest": 13.81,
      "sum": 3327.58,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 4400,
      "contrib": 100,
      "interest": 18.45,
      "sum": 4446.03,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 5500,
      "contrib": 100,
      "interest": 23.11,
      "sum": 5569.14,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 6600,
      "contrib": 100,
      "interest": 27.79,
      "sum": 6696.93,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 7700,
      "contrib": 100,
      "interest": 32.49,
      "sum": 7829.42,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 8800,
      "contrib": 100,
      "interest": 37.21,
      "sum": 8966.63,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 9900,
      "contrib": 100,
      "interest": 41.94,
      "sum": 10108.57,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 11000,
      "contrib": 100,
      "interest": 46.7,
      "sum": 11255.27,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 12100,
      "contrib": 100,
      "interest": 51.48,
      "sum": 12406.75,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 13200,
      "contrib": 100,
      "interest": 56.28,
      "sum": 13563.03,
      "tax": 0,
    },
  ],
  "sum": 13563.03,
  "taxSum": 0,
}
`;

exports[`interestjs node module. should count without params 1`] = `
Object {
  "capitalSum": 12000,
  "contributions": 0,
  "interestSum": 330.01,
  "payments": Array [
    Object {
      "capital": 1000,
      "capitalSum": 1000,
      "contrib": 0,
      "interest": 4.17,
      "sum": 1004.17,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 2000,
      "contrib": 0,
      "interest": 8.35,
      "sum": 2012.52,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 3000,
      "contrib": 0,
      "interest": 12.55,
      "sum": 3025.07,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 4000,
      "contrib": 0,
      "interest": 16.77,
      "sum": 4041.84,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 5000,
      "contrib": 0,
      "interest": 21.01,
      "sum": 5062.85,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 6000,
      "contrib": 0,
      "interest": 25.26,
      "sum": 6088.11,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 7000,
      "contrib": 0,
      "interest": 29.53,
      "sum": 7117.64,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 8000,
      "contrib": 0,
      "interest": 33.82,
      "sum": 8151.46,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 9000,
      "contrib": 0,
      "interest": 38.13,
      "sum": 9189.59,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 10000,
      "contrib": 0,
      "interest": 42.46,
      "sum": 10232.05,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 11000,
      "contrib": 0,
      "interest": 46.8,
      "sum": 11278.85,
      "tax": 0,
    },
    Object {
      "capital": 1000,
      "capitalSum": 12000,
      "contrib": 0,
      "interest": 51.16,
      "sum": 12330.01,
      "tax": 0,
    },
  ],
  "sum": 12330.01,
  "taxSum": 0,
}
`;
